src/arch/armv8/aarch32/inc/arch/spinlock.h

```diff
@@ -8,30 +8,62 @@
 
 #include <bao.h>
 
-typedef volatile uint32_t spinlock_t;
+typedef struct {
+    uint32_t ticket;
+    uint32_t next;
+} spinlock_t;
 
-#define SPINLOCK_INITVAL (0)
+#define SPINLOCK_INITVAL ((spinlock_t){0,0})
+
+static inline void spinlock_init(spinlock_t* lock)
+{
+    lock->ticket = 0;
+    lock->next = 0;
+}
+
+/**
+ * This lock follows the ticket lock algorithm described in Arm's ARM DDI0487I.a
+ * Appendix K13.
+ */
 
 static inline void spin_lock(spinlock_t* lock)
 {
-    spinlock_t const ONE = 1;
-    spinlock_t tmp;
+    uint32_t ticket;
+    uint32_t next;
+    uint32_t temp;
 
-    asm volatile(
+    (void)lock;
+    __asm__ volatile(
+        /* Get ticket */
         "1:\n\t"
-        "ldaex %r0, %1 \n\t"
-        "cmp %r0, #0 \n\t"
-        "strexeq %r0, %r2, %1 \n\t"
-        "cmpeq %r0, #0 \n\t"
+        "ldaex  %r0, %3\n\t"
+        "add    %r1, %r0, #1\n\t"
+        "dsb \n\t"
+        "strex  %r2, %r1, %3\n\t"
+        "cmp  %r2, #0\n\t"
         "bne 1b \n\t"
-        : "=&r"(tmp), "+Q"(*lock)
-        : "r"(ONE) : "memory");
+        /* Wait for your turn */
+        "2:\n\t"
+        "ldaex  %r1, %4\n\t"
+        "cmp    %r0, %r1\n\t"
+        "beq   3f\n\t"
+        "b 2b\n\t"
```

Comment:
is there a reason for not using wfe here?

Reply:
Added.